clients/admin-ui/src/features/custom-fields/CustomFieldModal.tsx

```diff
@@ -0,0 +1,293 @@
+    },
+  } as unknown as FormValues;
+};
+
+export const CustomFieldModal = ({
+  isOpen,
+  onClose,
+  isLoading,
+  customField,
+}: ModalProps) => {
+  const { errorAlert, successAlert } = useAlert();
+  const [addCustomFieldDefinition] = useAddCustomFieldDefinitionMutation();
+  const [updateCustomFieldDefinition] =
+    useUpdateCustomFieldDefinitionMutation();
+  const [upsertAllowList] = useUpsertAllowListMutation();
+
+  const { data: allowList, isLoading: isLoadingAllowList } =
+    useGetAllowListQuery(customField?.allow_list_id as string, {
+      skip: !customField?.allow_list_id,
+    });
+
+  const [validationSchema, setNewValidationSchema] = useState(
+    optionalValidationSchema
+  );
+
+  if (isLoadingAllowList || !isOpen) {
+    return null;
+  }
+
+  const transformedCustomField = transformCustomField(customField, allowList);
+  const initialValues = transformedCustomField || initialValuesTemplate;
+
+  const handelDropdownChange = (value: FieldTypes) => {
+    if (value === FieldTypes.OPEN_TEXT) {
+      setNewValidationSchema(optionalValidationSchema);
+    } else {
+      setNewValidationSchema(requiredValidationSchema);
+    }
+  };
+
+  // const test = useFormikContext<FormValues>()
+  // test.
+  const handleSubmit = async (
+    values: FormValues
+    // helpers: FormikHelpers<FormValues>
+  ) => {
+    if (
+      [FieldTypes.SINGLE_SELECT, FieldTypes.MULTIPLE_SELECT].includes(
+        values.field_type
+      )
+    ) {
+      const uniqueValues = new Set(
+        values.allow_list?.allowed_values
+          .map((v) => v.toLowerCase().trim())
+          .map((v) => v)
+      );
+      if (uniqueValues.size < values.allow_list!.allowed_values.length) {
+        errorAlert("List item value must be unique");
+        return;
+      }
+
+      const allowListPayload: AllowListUpdate = {
+        ...values.allow_list!,
+      };
+
+      if (values.allow_list_id) {
+        allowListPayload.id = values.allow_list_id;
+      }
+
+      if (!allowListPayload.name) {
+        // The UI no longer surfaces this field. It's a required field in the
+        // backend and must be unique. We generate a random name to avoid collisions.
+        allowListPayload.name =
+          Date.now().toString() + Math.random().toString();
+      }
+
+      const result = await upsertAllowList(allowListPayload);
+      if (!("error" in result) && !values.allow_list_id) {
+        // eslint-disable-next-line no-param-reassign
+        values.allow_list_id = result.data?.id;
+      }
+      // Add error handling here
+    }
+
+    if (values.field_type === FieldTypes.OPEN_TEXT) {
+      // eslint-disable-next-line no-param-reassign
+      values.allow_list_id = undefined;
+    }
+
+    if (
+      [FieldTypes.SINGLE_SELECT, FieldTypes.OPEN_TEXT].includes(
+        values.field_type
+      )
+    ) {
+      // eslint-disable-next-line no-param-reassign
+      values.field_type = AllowedTypes.STRING as unknown as FieldTypes;
```

Comment:
this is quite the typing tangle 😅 for things like this, it can be helpful to have another `transform` function which transforms from the form values back into the payload the endpoint is actually expecting. we don't have to address it now, but something to keep in mind

Reply:
it is... I'll keep this in mind in the future. I've seen those transform functions and now I understand why you use them 😅